src/mscorlib/shared/System/IO/FileStreamCompletionSource.Win32.cs

```diff
@@ -217,6 +217,29 @@ private static void Cancel(object state)
                     }
                 }
             }
+
+            public static FileStreamCompletionSource Create(FileStream stream, int numBufferedBytesRead, ReadOnlyMemory<byte> memory)
+            {
+                if (MemoryMarshal.TryGetArray(memory, out ArraySegment<byte> arraySegment))
+                {
+                    byte[] buffer = arraySegment.Array;
+
+                    // If the memory passed in isn't the stream's internal buffer, then delegate to
+                    // memory.Retain instead of passing the buffer to AllocateNativeOverlapped. In 
+                    // some cases, this will result in less pinning (in case the underlying memory is backed
+                    // by pre-pinned buffers)
+                    if (!ReferenceEquals(buffer, stream._buffer))
```

Comment:
We can remove the condition in the base class now, right? i.e. the code that's currently:
```C#
                _overlapped = (bytes == null || ReferenceEquals(bytes, _stream._buffer)) && _stream.CompareExchangeCurrentOverlappedOwner(this, null) == null ?
                    _stream._fileHandle.ThreadPoolBinding.AllocateNativeOverlapped(_stream._preallocatedOverlapped) :
                    _stream._fileHandle.ThreadPoolBinding.AllocateNativeOverlapped(s_ioCallback, this, bytes);
```
could instead be:
```
                Debug.Assert(bytes == null || ReferenceEquals(bytes, _stream._buffer));
                _overlapped = _stream.CompareExchangeCurrentOverlappedOwner(this, null) == null ?
                    _stream._fileHandle.ThreadPoolBinding.AllocateNativeOverlapped(_stream._preallocatedOverlapped) :
                    _stream._fileHandle.ThreadPoolBinding.AllocateNativeOverlapped(s_ioCallback, this, null);
```
?

Reply:
I'll leave it for the assert.